### DISCUSSION TITLE
Elliptic Curve Arithmetic Precompiles

### DISCUSSION THREAD

**Discussion by 0xpluto** [filecoin-project/FIPs#757] [Filecoin Virtual Machine]
### Simple Summary

This proposal will add helpful elliptic curve arithmetic operations to Filecoin's precompiles. This introduces modular exponentiation, elliptic curve addition, elliptic curve scalar multiplication, and bilinear function on groups for the elliptic curve 'alt_bn128'

### Motivation

Currently filecoin smart contracts are fully transparent. To protect privacy of users in certain applications, zkSNARKs must be utilized. Filecoin does not have any precompiles for these complex mathmatical operations. While some of these operations could fit into execution in plain code, gas costs would be unnecessarily high.

### Abstract

These EIP's suggest to add precompiled contracts for modular exponentiation, and addition, scalar multiplication, a pairing function on  a pairing-friendly elliptic curve 'alt_bn128'

[Precompiled contracts for addition and scalar multiplication on the elliptic curve alt_bn128](https://eips.ethereum.org/EIPS/eip-196)

[Precompiled contracts for optimal ate pairing check on the elliptic curve alt_bn128](https://eips.ethereum.org/EIPS/eip-197)

[Big integer modular exponentiation](https://eips.ethereum.org/EIPS/eip-198)

### Reference

These precompiles have specifics at [evm precompiles](evm.codes/precompiled) since the Byzantium hard fork.

Modular Exponentiation - 0x05

Elliptic Curve Addition - 0x06

Elliptic Curve Multiplication - 0x07

Elliptic Curve Pairing - 0x08

**Comment by Stebalien** [comment]
These precompiles already exist on the network (although they execute in WASM, not native).

**Reply by 0xpluto** [reply]
Are there bindings from the WASM implementation and the EVM?

**Reply by Stebalien** [reply]
They're _implemented_ and WASM and exposed in the EVM using the standard precompiles (0x05 - 0x08).

**Reply by 0xpluto** [reply]
Thank you, this helps. I was not aware they already existed.

**Reply by aashidham** [reply]
Is there a link to all the precompiles or some other resource to leave here for reference? Might be good for other readers of this thread.

**Reply by Stebalien** [reply]
They're the standard EVM precompiles 0x5 through 0x8 (links in the original post).
